README
I change json into an interactive form!


History.txt
== 0.1.3 2008-04-09
* Checkboxes automagically have a hidden field to make sure un-selected ones are sent.  Values are true / false

== 0.1.2 2008-03-21
* Fixed a bug where radio button groups didn't work in Firefox because of duplicate ids

== 0.1.1 2008-03-13
* Added a '+' delineator for any field that can be duplicated...
* Made selects only sort if they are a sub-select
* Fixed label bug for delineated fields

== 0.1.0 2008-03-05
* Form validation start and stop (...buggy)

== 0.0.9 2008-02-23

* A major refactor of the sub-select logic
* Support for all basic form elements


test/assets/test_helper.js
//I assume mootools was already required
String.implement('sexify', function(){
  return this.replace('test','').replace(/[A-Z]/g, function(match){
    return ' ' + match.charAt(0);
  })
})

DrNicTest.Unit.Testcase.prototype.$N = function(selector){
  return $$('#'+this.name+' '+selector)
}

var JSON = new Hash({

	encode: function(obj){
		switch ($type(obj)){
			case 'string':
				return '"' + obj.replace(/[\x00-\x1f\\"]/g, JSON.$replaceChars) + '"';
			case 'array':
				return '[' + String(obj.map(JSON.encode).filter($defined)) + ']';
			case 'object': case 'hash':
				var string = [];
				Hash.each(obj, function(value, key){
					var json = JSON.encode(value);
					if (json) string.push(JSON.encode(key) + ':' + json);
				});
				return '{' + String(string) + '}';
			case 'number': case 'boolean': return String(obj);
			case false: return 'null';
		}
		return null;
	},

	$specialChars: {'\b': '\\b', '\t': '\\t', '\n': '\\n', '\f': '\\f', '\r': '\\r', '"' : '\\"', '\\': '\\\\'},

	$replaceChars: function(chr){
		return JSON.$specialChars[chr] || '\\u00' + Math.floor(chr.charCodeAt() / 16).toString(16) + (chr.charCodeAt() % 16).toString(16);
	},

	decode: function(string, secure){
		if ($type(string) != 'string' || !string.length) return null;
		if (secure && !(/^[,:{}\[\]0-9.\-+Eaeflnr-u \n\r\t]*$/).test(string.replace(/\\./g, '@').replace(/"[^"\\\n\r]*"/g, ''))) return null;
		return eval('(' + string + ')');
	}

});

Native.implement([Hash, Array, String, Number], {

	toJSON: function(){
		return JSON.encode(this);
	}

});

src/base.js
//Globals...
Scroller = null
delay = 100
window.addEvent('load', function(){
  Scroller = new Fx.Scroll(window, {link:'chain', offset:{x:0,y:-25}})
})
JSAwesome = new Class({
	initialize: function(name, json, labels){
	  this.name = name
		this.json = json
		this.labels = labels || {}
		this.validatables = []
		this.validations = $H(labels).getKeys().filter(function(f){
		  return labels[f]['required'] || labels[f]['validation']
		})
	},
	to_html: function() {
	  var m = []
	  $A(this.json).each(function(p,i){
	    var n = []
	    //Allows you to reference previous selects...
	    if($type(p[1])=="number" && this.json[p[1]])
	      p[1] = this.json[p[1]][1]
	    if($type(p[0])=="array"){
	      n.push(p.map(function(r){
	        r = $splat(r)
	        n.push(this.extra('before', r[0]))
	        return new Element('div', {style:'float:left;margin-right:5px'}).adopt(this._process(r[1],r[0]))
	      }, this))
	      n.push(new Element('br', {style:'clear:left'}))
	      n.push(this.extra('after', $splat(p.getLast())[0]))
	    } else {
	      n.push(this.extra('before', p[0]))
	      n.push(this._process(p[1],p[0]))
	      n.push(this.extra('after', p[0]))
	    }
	    var klass = $type(p[0]) == "string" ? this._clean(p[0]) : "row_"+(i+1)
	    m.push(new Element('div', {'class':'error '+klass}).adopt(n))
	  }, this);
	  var adopted = $(this.name).adopt(m.concat($(this.name).getChildren().dispose()))
	  adopted.getElements('select').each(function(e){
	    e.fireEvent('change', {target:e})
	  })
	 return adopted
	},
	extra: function(where, what) {
	  return this.labels[where+'_'+what] ? new Element('div', {html: this.labels[where+'_'+what]}) : null
	},
	label: function(name, wafor) {
	  var newname = null
	  name = this._clean(name)
	  if($defined(this.labels[name])) {
	    if($defined(this.labels[name]['label']))
	      newname = this.labels[name]['label']
	    else
	      newname = this.labels[name]
	    if($type(newname) == "object")
	      newname = null
	  }
	  if(!$defined(newname))
	    newname = this._capitalize(name.replace(/_/g,' '))
	  return ($type(wafor) == "element" ? wafor.set('html', newname) : new Element('label', {'for':this._id(wafor), 'html':newname}))
	},
	addValidation: function(wha){
	  if(wha) {
	    this.validatables.each(function(v){
	      //There is an undefined element in IE for some sick reason
	      if(v && v.test(wha)) {
	        this.validatables.remove(v)
	        this.validations.push(v)
	      }
	    }, this)
	  } else {
	    this.validater = this.validate.bind(this)
	    $(this.name).getParent('form').addEvent('submit', this.validater)
	  }
	},
	stopValidation: function(wha){
	  if(wha) {
	    this.validations.each(function(v){
	      //There is an undefined element in IE for some sick reason
	      if(v && v.test(wha)) {
	        this.validations.remove(v)
	        this.validatables.push(v)
	      }
	    }, this)
	  } else
	    $(this.name).getParent('form').removeEvent('submit', this.validater)
	},
	validate: function(e) {
	  return this.validations.filter(function(r){
	    var checking = $(this.name).getElement('div .'+this._clean(r))
	    var error = checking.getParent('.error')
	    var invalid = this._check(checking)
	    if(invalid) {
	      this.delay = (delay -= 20)
	      invalid[1].set('style', 'background:#d88b7e')
	      if(error.getChildren().getLast().innerHTML != invalid[0]) {
	        var mes = new Element('span', {
	          style: 'color:red;display:block',
	          html:invalid[0]
	        }).inject(error)
	      }
	      invalid[1].addEvent(invalid[2] || 'blur', function(){
	        var check = this._check(checking)
	        if(!check || invalid[1] != check[1]) {
	          invalid[1].set('style', 'background:')
	          if(mes){
	            mes.dispose()
	            mes = false
	          }
	        }
	      }.bind(this))
	      if(this.delay > 0)
	        Scroller.toElement.delay(this.delay, Scroller, checking)
	      else
	        this._reset.delay(500)
	      return true
	    } else 
	      return false
	  }, this).length == 0
	},
	_reset: function() { delay = 100 },
	_check: function(element) {
	  var invalid = false
	  var label = this.labels[element.get('class')]
	  //Validate chained selects
	  if(element.get('tag') == 'select') {
      invalid = element.getParent().getElements("select, .custom").some(function(c){
        element = c
        return c.get('value') === ""
      });
      if(invalid) return ["This is a required field", element, 'change']
    } else if(element.get('type') == "radio" && label['required']) {
      var radios = element.getParent('fieldset').getElements('input')
      if(radios.some(function(r){return r.get('value')}))
        return false
      else
        return ["You must choose an option", new Elements(radios), 'click']
    } else if(element.get('type') == "checkbox" && label['required']) {
      if(element.checked)
        return false
      else
        return ["You must check this box", element, 'click']
    } else if(label['required'] && element.get('value') === "")
      return ["This is a required field", element]
    else if(label['validation']) {
      var args = $splat(label['validation'][0])
      var regex = new RegExp(args[0], args[1])
      return (regex.test(element.get('value')) ? false : [label['validation'][1], element])
    }
	},
	_process: function(cur, name, parent) {
	  var select_default = this.labels['{}'] || ["Choose Category", "Choose Subcategory"]
	  switch($type(cur)) {
	    //a select tag or group of radios/checkboxes
	    case 'array':
	      if(name.test(/^[*^]/)) {
	        return new Element('fieldset').adopt(
	          [this.label(this._clean(name), new Element('legend'))].concat(cur.map(function(c){
	            c = $splat(c)
	            return this._process(c[0], name, c.getLast());
	          }, this))
	        );
	      } else {          
          //Sort the options, and make the other / custom field go to the end... if it's a nested select
          if(parent) {
            cur = cur.sort()
            var other = false
            cur.some(function(a){ return a.test(/~/) ? other = a : other})
            if(other)
              cur.remove(other).push(other)
          }
          //Add the default header
          if(cur.length > 1)
	          cur = [parent ? select_default[1] : select_default[0]].concat(cur)
	        var level = parent ? parent.getParent().$attributes.level : null
	        var klass = [name,level].clean().join("_")
	        if(parent && cur.length > 0) {
	          parent.$attributes.extra = this._select(name, cur, klass, level)
	          return null
	        } else if(cur.length > 0)
	          return [this.label(name, name),this._select(name, cur, klass, level)]
	        else return null
	      }
	    case 'object':
	      cur = $H(cur)
	      var t = [this._process(cur.getKeys(), name, parent)]
	      var root = parent ? parent.$attributes.extra : t[0][1]
	      t.push(this._process([select_default[1]], name, root.getElement('option')))
	      return t.concat(cur.getValues().map(function(v){
	        var val = cur.keyOf(v).split('|').getLast()
	        var parent = root.getElement('option[value='+val+']')
	        return this._process(v, name, parent)
	      }, this));
	    default:
	      var e = null
        //Ugggggg
	      if($type(cur)=="string") {
	        //There should be a better way to do this
	        var reversed = [].concat(select_default).reverse()
	        var val = cur.split('|').getLast().replace(new RegExp(reversed.join("|")),'')
	        cur = this._clean(cur.split('|')[0])
	      }
	      if($type(parent) == "element") {
	        //Allows for custom values
	        e = new Element('option', {html: cur, value: this._clean(val)})
	        if(val.test(/^~/))
            e.$attributes.extra = this._custom(parent.get('name').replace(/(.+)\[(.+)\]/, '$1[$2_other]'), parent.get('class'))
	      } else {
	        if(name.test(/^#/)) {
	          e = new Element('textarea', {
	            name: this._name(name),
	            id: this._id(name),
	            'class':this._clean(name), 
	            html: cur})
	        } else if(name.test(/^_/)) {
	          e = this._input('hidden', name, cur)
	        } else if(name.test(/^\*/)) {
            e = this.label(cur, name).grab(
              this._input('radio', name, val), 'top')
          } else if(name.test(/^\^/)) {
            //handles both grouped checkboxes and individual ones... Kindof a Mind FFFF
            if($defined(parent))
              var tname = cur
            else {
              parent = cur
              var tname = name
            }
            e = this.label(tname, val || tname).grab(
              this._input('checkbox', (val || tname), parent === true), 'top')
	        } else e = this._input('text', name, cur)
	        //For duplication...
	        if(name.test(/^.?\+/))
	          e = [e, new Element('a', {html:' +', href:'#', events:{click:this.duplicate}, 'class':'plus'})]
	        if(!name.test(/^[_*^]/))
	          e = [this.label(name, name), e]
	      }
	      return e
	  }
	},
	duplicate: function(e) {
	  e.stop()
	  var orig = e.target.getPrevious()
	  if(!orig.name.test(/\[\]$/))
	    orig.name = orig.name+"[]"
	  orig.clone().inject(orig, 'after')
	  new Element('br').inject(orig, 'after')
	},
	_capitalize: function(string){
		return string.replace(/^[a-z]/, function(match){ return match.toUpperCase() });
	},
	_clean: function(name) {
	  return name.replace(/^[_#*^~]?\+?/,'')
	},
	_name: function(name) {
	  return this.name+'['+this._clean(name)+']'
	},
	_id: function(name) {
	  return this.name+'_'+this._clean(name)
	},
	_input: function(type, name, val) {
	  var e = new Element('input', {type: type, 'class': this._clean(name), name: this._name(name), id: (type =='radio' ? '' : this._id(name))})
    if(type == "checkbox") {
      e.set('checked', val ? "checked" : "")
      e.set('value', 'true')
      e = new Element('span').adopt([e, this._input('hidden', name, 'false')])
    } else
      e.set('value', val)
    return e
	},
	_custom: function(name, klass) {
	  var val = this.labels['~'] || "Custom..."
	  return new Element('input', {
      'class':klass+' custom', 
      type: 'text', 
      name: name, 
      value: val,
      events: {
        'focus': function(){
          if(this.value == val)
            this.set('value', "")
        }
      }
    })
	},
	_select: function(name, options, klass, level) {
	  var level = level || 0
	  var leveled = level > 0 ? name + '_' + level : name
	  var select = new Element("select", {
      name: this._name(leveled),
      id: this._id(leveled),
      'class': klass,
      events: {
        change: function(event){
          var e = event.target
          var option = e[e.selectedIndex]
          var klass = e.get('class').split(' ')[0]
          var verify = new RegExp("("+name+')_\\d+$')
          var next = klass.test(verify) ? klass.replace(verify, '$1_'+(level+1)) : klass+'_'+(level+1) 
          //Dispose namespaced in a wrapper
          $E('#'+this.name+' .'+klass.replace(verify,'$1')).getElements('.custom, select').each(function(i){
            if(i.hasClass('custom') || i.$attributes.level > level+1)
              i.dispose()
          })
          if(!option.$attributes.extra) return
          var test = $(this.name).getElement('.'+next)
          //Replace or add the extras
          if(test)
            var made = option.$attributes.extra.replaces(test)
          else
            var made = option.$attributes.extra.inject(e, "after")
          made.fireEvent('change', {target:made})
        }.bind(this)
      }
    })    
    select.$attributes.level = (level || 0) + 1
    return select.adopt(options.map(
      function(o){
        return this._process(o, klass, select)
      },this)
    )
	}
});

js/jsawesome_pre.js
//Globals...
Scroller = null
delay = 100
window.addEvent('load', function(){
  Scroller = new Fx.Scroll(window, {link:'chain', offset:{x:0,y:-25}})
})
JSAwesome = new Class({
	initialize: function(name, json, labels){
	  this.name = name
		this.json = json
		this.labels = labels || {}
		this.validations = $H(labels).getKeys().filter(function(f){
		  return labels[f]['required'] || labels[f]['validation']
		})
		this.level = 0 
		this.in_select = false
		this.nested = {}
	},
	to_html: function() {
	  var m = []
	  $A(this.json).each(function(p){
	    var n = []
	    //Allows you to reference previous selects...
	    if($type(p[1])=="number" && this.json[p[1]])
	      p[1] = this.json[p[1]][1]
	    if($type(p[0])=="array"){
	      n.push(p.map(function(r){
	        return new Element('div', {style:'float:left;margin-right:5px'}).adopt(this._process(r[1],r[0]))
	      }, this))
	      n.push(new Element('br', {style:'clear:left'}))
	    } else
	      n.push(this._process(p[1],p[0]))
	    m.push(new Element('div', {'class':'error '+p[0]}).adopt(n))
	    this.level = 0
	  }, this);
	 return $(this.name).adopt(m)
	},
	label: function(name, wafor) {
	  var newname = null
	  if($defined(this.labels[name])) {
	    if($defined(this.labels[name]['label']))
	      newname = this.labels[name]['label']
	    else
	      newname = this.labels[name]
	    if($type(newname) == "object")
	      newname = null
	  }
	  if(!$defined(newname))
	    newname = name.replace(/^[_#*^]/,'').replace(/_/g,' ').capitalize()
	  return ($type(wafor) == "element" ? wafor.set('html', newname) : new Element('label', {'for':wafor, 'html':newname}))
	},
	validate: function(e) {
	  return this.validations.every(function(r){
	    var checking = $(this.name).getElement('.'+r.replace(/^[_#~*^]/,''))
	    var error = checking.getParent('.error')
	    var invalid = this._check(checking)
	    if(invalid) {
	      this.delay = (delay -= 20)
	      invalid[1].set('style', 'background:#d88b7e')
	      if(error.getChildren().getLast().innerHTML != invalid[0]) {
	        var mes = new Element('div', {
	          style: 'color:red',
	          html:invalid[0]
	        }).inject(error)
	      }
	      invalid[1].addEvent(invalid[2] || 'blur', function(){
	        var check = this._check(checking)
	        if(!check || invalid[1] != check[1]) {
	          invalid[1].set('style', 'background:')
	          if(mes){
	            mes.dispose()
	            mes = false
	          }
	        }
	      }.bind(this))
	      if(this.delay > 0)
	        Scroller.toElement.delay(this.delay, Scroller, checking)
	      else
	        this._reset.delay(500)
	      return false
	    } else 
	      return true
	  }, this)
	},
	_reset: function() { delay = 100 },
	_check: function(element) {
	  var invalid = false
	  var label = this.labels[element.get('class')]
	  if(element.get('tag') == 'div') {
      invalid = !element.getChildren().every(function(c){
        element = c
        return !(c.get('value') === "")
      });
      if(invalid) return ["This is a required field", element, 'change']
    } else if(element.get('type') == "radio" && label['required']) {
      var radios = element.getParent('fieldset').getElements('input')
      if(radios.some(function(r){return r.get('value')}))
        return false
      else
        return ["You must choose an option", new Elements(radios), 'click']
    } else if(label['required'] && element.get('value') === "")
      return ["This is a required field", element]
    else if(label['validation'] && element.get('value') !== "") {
      var args = $splat(label['validation'][0])
      var regex = new RegExp(args[0], args[1])
      return (regex.test(element.get('value')) ? false : [label['validation'][1], element])
    }
	},
	_process: function(cur, names, nested) {
	  names = $splat(names);
	  switch($type(cur)) {
	    //a select tag
	    case 'array':
	      if(names[0].test(/^\*/)) {
	        return new Element('fieldset').adopt(
	        [this.label(names[0].substring(1), new Element('legend'))].concat(cur.map(function(c){
	          return this._process(c, names);
	        }, this)));
	      } else {
	        this.in_select = true
	        var root = !this.level
	        cur = cur.sort()
	        //Make the other / custom field go to the end...
	        var other = false
	        cur.some(function(a){
	          if(a.test(/~/))
	            other = a
	          return other
	        })
	        if(other)
	          cur.remove(other).push(other)
	        if(cur.length > 1)
	          cur = ["Choose "+(this.level==0 ? "Category" : "Subcategory")].concat(cur)
	        else if(this.level <= 1 && cur.length > 0)
	          root = true
	        var name = this.name+"_"+names[0]
	        var klass = [names[0],this.level].join("_")
	        var els = []
	        if(!root) {
	          //Add the subcategory dropdown
	          var store = klass
	          if(names.getLast() == "_") {
	            var classes = klass.split('_')
	            //Go back one level
	            store = classes.concat([classes.pop().toInt() - 1]).join('_')
	            names.pop()
	          }
	          if(cur.length > 0) this._store(store, names.getLast(), cur, klass, name)
	          this.in_select = false
	          return null
	        } else {
	          if(this.level==0)
	            els.push(this.label(names[0], name))
	          if(!nested)
	            klass += ' '+names[0]
	          els = els.concat(this._nested(name, cur, klass))
	          this.in_select = false
  	        return els
	        }
	      }
	    case 'object':
	      cur = $H(cur)
	      var t = [this._process(cur.getKeys(), names, true)]
	      this.level += 1
	      t.push(this._process(["Choose Subcategory"], names.concat(['_']), true))
	      return t.concat(cur.getValues().map(function(v){
	        return this._process(v, names.concat([cur.keyOf(v).split('|').getLast()]), true)
	      }, this));
	    default:
	      var name = names.join("_");
	      var e = null
	      //AKA not a checkbox
	      if($type(cur)=="string") {
	        var val = cur.split('|').getLast().replace(/Choose (Sub)?Category/i,'')
	        cur = cur.split('|')[0].replace(/^[~*]/,'')
	      }
	      if(this.in_select) {
	        //Allows for custom values
	        e = new Element('option', {html: cur, value: val})
	        //Wow this is sketchy as F
	        var name = name.split(' ')[0]
	        if(val.test(/^~/)) this._store(name, val, [], name, this.name+'_'+name)
	      } else {
	        if(name.test(/^#/)) {
	          e = new Element('textarea', {
	            name: this.name+'_'+name.substring(1), 
	            'class':name.substring(1), 
	            html: cur})
	        } else if(name.test(/^_/)) {
	          e = new Element('input', {
	            type: 'hidden',
	            'class': name.substring(1),
	            name: this.name+'_'+name.substring(1), 
	            value: cur
	          })
	        } else if(name.test(/^\*/)) {
            e = this.label(cur, this.name+'_'+name.substring(1)).grab(new Element('input', {
              type: 'radio',
              'class': name.substring(1),
              name: this.name+'_'+name.substring(1),
              value: val
            }), 'top')
          } else if(name.test(/^\^/)) {
            e = this.label(name, this.name+'_'+name.substring(1)).grab(new Element('input', {
              type: 'checkbox',
              'class': name.substring(1),
              name: this.name+'_'+name.substring(1),
              checked: (cur ? "checked" : "")
            }), 'top')
	        } else e = new Element('input', {type: 'text', 'class': name, name: this.name+'_'+name, value: cur})
	        if(!name.test(/^[_*^]/))
	          e = [this.label(name, e.name), e]
	      }
	      return e
	  }
	},
	_custom: function(name, klass) {
	  var val = this.labels['~'] || "Custom..."
	  return new Element('input', {
      'class':(this.level > 1 ? klass+' sub' : klass)+' custom', 
      type: 'text', 
      name: name, 
      value: val,
      events: {
        'focus': function(){
          if(this.value == val)
            this.set('value', "")
        }
      }
    })
	},
	_store: function(store, key, options, klass, name) {
	  if(key.test(/^~/)){
      var type = "custom"
      name = name+'_other'
    }else{
      name = name+'_sub'+this.level
      var type = options.length == 0 ? "disabled" : "select"
    }
    //if(this.nested[store] && this.nested[store][key]) return
	  this.nested[store] = this.nested[store] || {}
    this.nested[store][key] = this.nested[store][key] || []
    this.nested[store][key].push(this._nested(name, options, klass, type))
	},
	_nested: function(name, options, klass, type) {
	  type = type || 'select'
	  if(type == "custom") {
	    return this._custom(name, klass)
	  } else {
	    var select = new Element("select", {
        name: name,
        'class': this.level > 1 ? klass+' sub' : klass,
        events: {
          change: function(event){
            var e = event.target
            var classes = e.get('class').split('_')
            var level = classes.pop().toInt()+1
            var klass = classes.concat([level]).join('_')
            var it = e.get('value')
            //Dispose namespaced in a wrapper
            if(level < 3)
              $E('#'+this.name+' .'+classes.join('_')).getElements('.custom, .sub').dispose()
            console.log('Checking for', klass, 'or', e.get('class').split(' ')[0])
            var there = this.nested[klass] || this.nested[e.get('class').split(' ')[0]];
            if(!there) return
            var child = there[it]
            if(!child) return
            var test = $(this.name).getElement('.'+klass)
            if(test)
              var made = child[0].replaces(test)
            else
              var made = child[0].inject(e.getParent())
            //Restore the old value instead of an empty select
            if(child.length > 1) {
              if(made.selectedIndex > 0)                
                var child = this.nested[classes.concat([level+1]).join('_')][made.get('value')]
              child.getLast().inject(made, 'after')
            }
            //$E('#'+this.name+' .'+classes.join('_')).getElements('select').fireEvent('change')      
          }.bind(this)
        }
      })
      if(type == "disabled")
        select.set('style', 'display:none').set('disabled', 'disabled')
      return select.adopt(options.map(
        function(o){
          return this._process(o, klass)
        },this)
      )
    }
	}
});

/*
{'text':'',
  '#textarea':'Default text',
  '_hidden':'invisible', 
  {'sub_cats': 
    {'rad': ['cool', 'neat'], 
     'awesome': 
       {'crazy': ['indeed', 'man'],
        'way': ['oh', 'yeah']}
    }
  },
  {'drop': ['single', 'dude']}} =>
  <input type="text" name="text" value=""/>
  <textarea name="textarea">Default text</textarea>
  <input type="hidden" name="hidden" value="invisible"/>
  <select name="sub_cats" class="sub_cats_0">
    <option>Choose Category</option>
    <option>rad</option>
    <option>awesome</option>
  </select>
  <select name="sub_cats_0" class="sub_cats_1">
    <option>Choose Subcategory</option>
  </select>
  <select name="sub_cats_rad" class="sub_cats_1" style="display:none">
    <option>cool</option>
    <option>neat</option>
  </select>
  <select name="sub_cats_awesome" class="sub_cats_1" style="display:none">
    <option>crazy</option>
    <option>way</option>
  </select>
  <select name="sub_cats_awesome_0" class="sub_cats_2" style="display:none">
    <option>Choose Subcategory</option>
  </select>
  <select name="sub_cats_awesome_crazy" class="sub_cats_2" style="display:none">
    <option>indeed</option>
    <option>man</option>
  </select>
  <select name="sub_cats_awesome_way" class="sub_cats_2" style="display:none">
    <option>oh</option>
    <option>yeah</option>
  </select>
  <select name="drop">
    <option>single</option>
    <option>dude</option>
  </select>
  */